File: client/src/services/messageTemplateService.ts
import { apiClient } from '../api/client';

export interface CreateMessageTemplateRequest {
  companyId: string;
  name: string;
  content: string;
  isAiGenerated: boolean;
  aiAgentId?: string;
  tone?: string;
}

export interface UpdateMessageTemplateRequest {
  name?: string;
  content?: string;
  tone?: string;
}

export interface MessageTemplateResponse {
  id: string;
  companyId: string;
  companyName: string;
  name: string;
  content: string;
  isAiGenerated: boolean;
  createdByUserId?: string;
  createdByUserName?: string;
  aiAgentId?: string;
  aiAgentName?: string;
  tone?: string;
  lastEditedByUserId?: string;
  lastEditedByUserName?: string;
  editCount: number;
  createdAt: string;
  updatedAt: string;
}

export const messageTemplateService = {
  async create(data: CreateMessageTemplateRequest): Promise<MessageTemplateResponse> {
    try {
      const response = await apiClient.post<MessageTemplateResponse>('/api/message-templates', data);
      return response;
    } catch (error) {
      console.error('Error creating message template:', error);
      throw error;
    }
  },

  async getAll(): Promise<MessageTemplateResponse[]> {
    try {
      const response = await apiClient.get<MessageTemplateResponse[]>('/api/message-templates');
      return response;
    } catch (error) {
      console.error('Error fetching message templates:', error);
      throw error;
    }
  },

  async getById(id: string): Promise<MessageTemplateResponse> {
    try {
      const response = await apiClient.get<MessageTemplateResponse>(`/api/message-templates/${id}`);
      return response;
    } catch (error) {
      console.error('Error fetching message template:', error);
      throw error;
    }
  },

  async update(id: string, data: UpdateMessageTemplateRequest): Promise<MessageTemplateResponse> {
    try {
      const response = await apiClient.put<MessageTemplateResponse>(`/api/message-templates/${id}`, data);
      return response;
    } catch (error) {
      console.error('Error updating message template:', error);
      throw error;
    }
  },

  async delete(id: string): Promise<void> {
    try {
      await apiClient.delete(`/api/message-templates/${id}`);
    } catch (error) {
      console.error('Error deleting message template:', error);
      throw error;
    }
  },

  async getByCompany(companyId: string): Promise<MessageTemplateResponse[]> {
    try {
      const response = await apiClient.get<MessageTemplateResponse[]>(`/api/message-templates/company/${companyId}`);
      return response;
    } catch (error) {
      console.error('Error fetching company message templates:', error);
      throw error;
    }
  },

  async search(term: string): Promise<MessageTemplateResponse[]> {
    try {
      const response = await apiClient.get<MessageTemplateResponse[]>(`/api/message-templates/search?term=${encodeURIComponent(term)}`);
      return response;
    } catch (error) {
      console.error('Error searching message templates:', error);
      throw error;
    }
  },

  async getRevisionHistory(templateId: string): Promise<MessageTemplateRevision[]> {
    try {
      const response = await apiClient.get<MessageTemplateRevision[]>(`/api/message-template-revisions/template/${templateId}`);
      return response;
    } catch (error) {
      console.error('Error fetching template revision history:', error);
      throw error;
    }
  },

  async getDeleted(): Promise<MessageTemplateResponse[]> {
    try {
      const response = await apiClient.get<MessageTemplateResponse[]>('/api/message-templates', { 
        deleted: 'true' 
      });
      return response;
    } catch (error) {
      console.error('Error fetching deleted message templates:', error);
      throw error;
    }
  },

  async restore(id: string): Promise<MessageTemplateResponse> {
    try {
      const response = await apiClient.post<MessageTemplateResponse>(`/api/message-templates/${id}/restore`);
      return response;
    } catch (error) {
      console.error('Error restoring message template:', error);
      throw error;
    }
  }
};

export type RevisionType = 'CREATE' | 'EDIT' | 'DELETE' | 'RESTORE';

export interface MessageTemplateRevision {
  id: string;
  templateId: string;
  revisionNumber: number;
  content: string;
  editedByUserId?: string;
  editedByUserName?: string;
  revisionType: RevisionType;
  revisionTimestamp: string;
  createdAt: string;
  updatedAt: string;
}

File: api-dotnet/Rubia.Server/Services/MessageTemplateService.cs
using Microsoft.EntityFrameworkCore;
using Rubia.Server.Data;
using Rubia.Server.DTOs;
using Rubia.Server.Entities;
using Rubia.Server.Services.Interfaces;

namespace Rubia.Server.Services;

public class MessageTemplateService : IMessageTemplateService
{
    private readonly RubiaDbContext _context;
    private readonly ILogger<MessageTemplateService> _logger;

    public MessageTemplateService(RubiaDbContext context, ILogger<MessageTemplateService> logger)
    {
        _context = context;
        _logger = logger;
    }

    public async Task<MessageTemplateDto> CreateAsync(CreateMessageTemplateDto createDto, Guid? currentUserId = null)
    {
        _logger.LogInformation("Creating message template: {Name} for company: {CompanyId}", createDto.Name, createDto.CompanyId);

        // Validate company exists
        var company = await _context.Companies.FirstOrDefaultAsync(c => c.Id == createDto.CompanyId);
        if (company == null)
        {
            throw new ArgumentException("Empresa não encontrada");
        }

        // Check if template name already exists for this company
        if (await ExistsByNameAndCompanyAsync(createDto.Name, createDto.CompanyId))
        {
            throw new ArgumentException($"Template com nome '{createDto.Name}' já existe nesta empresa");
        }

        // Validate AI Agent if provided
        if (createDto.AIAgentId.HasValue)
        {
            var aiAgent = await _context.AIAgents.FirstOrDefaultAsync(a => a.Id == createDto.AIAgentId.Value && a.CompanyId == createDto.CompanyId);
            if (aiAgent == null)
            {
                throw new ArgumentException("Agente de IA não encontrado ou não pertence a esta empresa");
            }
        }

        var template = new MessageTemplate
        {
            CompanyId = createDto.CompanyId,
            Name = createDto.Name,
            Content = createDto.Content,
            IsAIGenerated = createDto.IsAIGenerated,
            AIAgentId = createDto.AIAgentId,
            Tone = createDto.Tone?.ToUpper(),
            CreatedByUserId = currentUserId,
            EditCount = 0
        };

        _context.MessageTemplates.Add(template);
        await _context.SaveChangesAsync();

        _logger.LogInformation("Message template created successfully with id: {Id}", template.Id);
        return await GetByIdAsync(template.Id, createDto.CompanyId);
    }

    public async Task<MessageTemplateDto> GetByIdAsync(Guid id, Guid companyId)
    {
        _logger.LogDebug("Fetching message template with id: {Id} for company: {CompanyId}", id, companyId);

        var template = await _context.MessageTemplates
            .Include(t => t.Company)
            .Include(t => t.CreatedByUser)
            .Include(t => t.AIAgent)
            .Include(t => t.LastEditedByUser)
            .FirstOrDefaultAsync(t => t.Id == id && t.CompanyId == companyId && t.DeletedAt == null);

        if (template == null)
        {
            throw new ArgumentException("Template de mensagem não encontrado");
        }

        return ToDto(template);
    }

    public async Task<List<MessageTemplateDto>> GetAllByCompanyAsync(Guid companyId)
    {
        _logger.LogDebug("Fetching all message templates for company: {CompanyId}", companyId);

        var templates = await _context.MessageTemplates
            .Include(t => t.Company)
            .Include(t => t.CreatedByUser)
            .Include(t => t.AIAgent)
            .Include(t => t.LastEditedByUser)
            .Where(t => t.CompanyId == companyId)
            .OrderBy(t => t.Name)
            .ToListAsync();

        return templates.Select(ToDto).ToList();
    }

    public async Task<List<MessageTemplateDto>> GetActiveByCompanyAsync(Guid companyId)
    {
        _logger.LogDebug("Fetching active message templates for company: {CompanyId}", companyId);

        var templates = await _context.MessageTemplates
            .Include(t => t.Company)
            .Include(t => t.CreatedByUser)
            .Include(t => t.AIAgent)
            .Include(t => t.LastEditedByUser)
            .Where(t => t.CompanyId == companyId && t.DeletedAt == null)
            .OrderBy(t => t.Name)
            .ToListAsync();

        return templates.Select(ToDto).ToList();
    }

    public async Task<List<MessageTemplateDto>> GetByCreatedByUserAsync(Guid userId, Guid companyId)
    {
        _logger.LogDebug("Fetching message templates created by user: {UserId} for company: {CompanyId}", userId, companyId);

        var templates = await _context.MessageTemplates
            .Include(t => t.Company)
            .Include(t => t.CreatedByUser)
            .Include(t => t.AIAgent)
            .Include(t => t.LastEditedByUser)
            .Where(t => t.CompanyId == companyId && t.CreatedByUserId == userId && t.DeletedAt == null)
            .OrderBy(t => t.Name)
            .ToListAsync();

        return templates.Select(ToDto).ToList();
    }

    public async Task<List<MessageTemplateDto>> GetByAIAgentAsync(Guid aiAgentId, Guid companyId)
    {
        _logger.LogDebug("Fetching message templates by AI agent: {AIAgentId} for company: {CompanyId}", aiAgentId, companyId);

        var templates = await _context.MessageTemplates
            .Include(t => t.Company)
            .Include(t => t.CreatedByUser)
            .Include(t => t.AIAgent)
            .Include(t => t.LastEditedByUser)
            .Where(t => t.CompanyId == companyId && t.AIAgentId == aiAgentId && t.DeletedAt == null)
            .OrderBy(t => t.Name)
            .ToListAsync();

        return templates.Select(ToDto).ToList();
    }

    public async Task<List<MessageTemplateDto>> GetByToneAsync(string tone, Guid companyId)
    {
        _logger.LogDebug("Fetching message templates by tone: {Tone} for company: {CompanyId}", tone, companyId);

        var templates = await _context.MessageTemplates
            .Include(t => t.Company)
            .Include(t => t.CreatedByUser)
            .Include(t => t.AIAgent)
            .Include(t => t.LastEditedByUser)
            .Where(t => t.CompanyId == companyId && t.Tone == tone.ToUpper() && t.DeletedAt == null)
            .OrderBy(t => t.Name)
            .ToListAsync();

        return templates.Select(ToDto).ToList();
    }

    public async Task<List<MessageTemplateDto>> GetAIGeneratedByCompanyAsync(Guid companyId)
    {
        _logger.LogDebug("Fetching AI-generated message templates for company: {CompanyId}", companyId);

        var templates = await _context.MessageTemplates
            .Include(t => t.Company)
            .Include(t => t.CreatedByUser)
            .Include(t => t.AIAgent)
            .Include(t => t.LastEditedByUser)
            .Where(t => t.CompanyId == companyId && t.IsAIGenerated && t.DeletedAt == null)
            .OrderBy(t => t.Name)
            .ToListAsync();

        return templates.Select(ToDto).ToList();
    }

    public async Task<List<MessageTemplateDto>> GetManualByCompanyAsync(Guid companyId)
    {
        _logger.LogDebug("Fetching manual message templates for company: {CompanyId}", companyId);

        var templates = await _context.MessageTemplates
            .Include(t => t.Company)
            .Include(t => t.CreatedByUser)
            .Include(t => t.AIAgent)
            .Include(t => t.LastEditedByUser)
            .Where(t => t.CompanyId == companyId && !t.IsAIGenerated && t.DeletedAt == null)
            .OrderBy(t => t.Name)
            .ToListAsync();

        return templates.Select(ToDto).ToList();
    }

    public async Task<MessageTemplateDto> UpdateAsync(Guid id, UpdateMessageTemplateDto updateDto, Guid companyId, Guid? currentUserId = null)
    {
        _logger.LogInformation("Updating message template with id: {Id} for company: {CompanyId}", id, companyId);

        var template = await _context.MessageTemplates
            .Include(t => t.Company)
            .Include(t => t.CreatedByUser)
            .Include(t => t.AIAgent)
            .Include(t => t.LastEditedByUser)
            .FirstOrDefaultAsync(t => t.Id == id && t.CompanyId == companyId && t.DeletedAt == null);

        if (template == null)
        {
            throw new ArgumentException("Template de mensagem não encontrado");
        }

        if (updateDto.Name != null && updateDto.Name != template.Name)
        {
            if (await ExistsByNameAndCompanyAsync(updateDto.Name, companyId))
            {
                throw new ArgumentException($"Template com nome '{updateDto.Name}' já existe nesta empresa");
            }
            template.Name = updateDto.Name;
        }

        if (updateDto.Content != null) template.Content = updateDto.Content;
        if (updateDto.Tone != null) template.Tone = updateDto.Tone.ToUpper();

        template.LastEditedByUserId = currentUserId;
        template.EditCount++;

        await _context.SaveChangesAsync();
        _logger.LogInformation("Message template updated successfully");

        // Reload to get updated related entities
        await _context.Entry(template).ReloadAsync();
        await _context.Entry(template).Reference(t => t.Company).LoadAsync();
        await _context.Entry(template).Reference(t => t.CreatedByUser).LoadAsync();
        await _context.Entry(template).Reference(t => t.AIAgent).LoadAsync();
        await _context.Entry(template).Reference(t => t.LastEditedByUser).LoadAsync();

        return ToDto(template);
    }

    public async Task SoftDeleteAsync(Guid id, Guid companyId)
    {
        _logger.LogInformation("Soft deleting message template with id: {Id} for company: {CompanyId}", id, companyId);

        var template = await _context.MessageTemplates
            .FirstOrDefaultAsync(t => t.Id == id && t.CompanyId == companyId && t.DeletedAt == null);

        if (template == null)
        {
            throw new ArgumentException("Template de mensagem não encontrado");
        }

        template.DeletedAt = DateTime.UtcNow;
        await _context.SaveChangesAsync();

        _logger.LogInformation("Message template soft deleted successfully");
    }

    public async Task DeleteAsync(Guid id, Guid companyId)
    {
        _logger.LogInformation("Deleting message template with id: {Id} for company: {CompanyId}", id, companyId);

        var template = await _context.MessageTemplates
            .FirstOrDefaultAsync(t => t.Id == id && t.CompanyId == companyId);

        if (template == null)
        {
            throw new ArgumentException("Template de mensagem não encontrado");
        }

        _context.MessageTemplates.Remove(template);
        await _context.SaveChangesAsync();

        _logger.LogInformation("Message template deleted successfully");
    }

    public async Task<long> CountByCompanyAsync(Guid companyId)
    {
        return await _context.MessageTemplates
            .Where(t => t.CompanyId == companyId)
            .CountAsync();
    }

    public async Task<long> CountActiveByCompanyAsync(Guid companyId)
    {
        return await _context.MessageTemplates
            .Where(t => t.CompanyId == companyId && t.DeletedAt == null)
            .CountAsync();
    }

    public async Task<long> CountAIGeneratedByCompanyAsync(Guid companyId)
    {
        return await _context.MessageTemplates
            .Where(t => t.CompanyId == companyId && t.IsAIGenerated && t.DeletedAt == null)
            .CountAsync();
    }

    public async Task<bool> ExistsByNameAndCompanyAsync(string name, Guid companyId)
    {
        return await _context.MessageTemplates
            .AnyAsync(t => t.Name == name && t.CompanyId == companyId && t.DeletedAt == null);
    }

    private static MessageTemplateDto ToDto(MessageTemplate template)
    {
        return new MessageTemplateDto
        {
            Id = template.Id,
            CompanyId = template.CompanyId,
            CompanyName = template.Company?.Name ?? string.Empty,
            Name = template.Name,
            Content = template.Content,
            IsAIGenerated = template.IsAIGenerated,
            CreatedByUserId = template.CreatedByUserId,
            CreatedByUserName = template.CreatedByUser?.Name ?? string.Empty,
            AIAgentId = template.AIAgentId,
            AIAgentName = template.AIAgent?.Name ?? string.Empty,
            Tone = template.Tone,
            LastEditedByUserId = template.LastEditedByUserId,
            LastEditedByUserName = template.LastEditedByUser?.Name ?? string.Empty,
            EditCount = template.EditCount,
            CreatedAt = template.CreatedAt,
            UpdatedAt = template.UpdatedAt,
            DeletedAt = template.DeletedAt
        };
    }
}